### PR TITLE
Keep the `epi_df` class when head and tail are used

### PR DESCRIPTION
**Pull request by rachlobay**
As per [issue 96](https://github.com/cmu-delphi/epiprocess/issues/96) `head.epi_df` and `tail.epi_df` to keep the `epi_df` class when they are used. Note the tests for this in test-methods-epi_df.R

### CODE REVIEW

File: tests/testthat/test-methods-epi_df.R [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      length.out = 5
+    ), times = 2),
+    geo_value = rep(c("ca", "hi"), each = 5)
+  ) %>% epiprocess::as_epi_df()
+  
+  expect_true(is_epi_df(head(tib)))
+  expect_true(is_epi_df(tail(tib)))
+  expect_identical(attributes(head(tib))$metadata$geo_type, attributes(tib)$metadata$geo_type)
+  expect_identical(attributes(head(tib))$metadata$time_type, attributes(tib)$metadata$time_type)
+  expect_identical(attributes(head(tib))$metadata$as_of, attributes(tib)$metadata$as_of)
```
Comment (brookslogan) [inline tests/testthat/test-methods-epi_df.R:17]: Is `additional_metadata` also supposed to be preserved?  The tests for head and tail here should probably ensure that the desired behavior (retaining, dropping, etc.) is achieved.

File: R/methods-epi_df.R [hunk 2]
```diff
@@ -67,15 +67,15 @@ summary.epi_df = function(object, ...) {
 #' @export
 #' @noRd
 head.epi_df = function(x, ...) {
-  head(tibble::as_tibble(x), ...)
+  as_epi_df(NextMethod("head"))
```
Comment (brookslogan) [inline R/methods-epi_df.R:70]: Minor "style"-like thing: might want to exclude `"head"` if it's not necessary to match how `NextMethod` is called elsewhere in the file, or explain why it's necessary in a comment if it is necessary.

**Comment by brookslogan** [comment]
The comments above are very minor.  This PR could be merged to address #96 without any changes.

File: tests/testthat/test-methods-epi_df.R [hunk 3]
```diff
@@ -0,0 +1,21 @@
+test_that("head and tail do not drop the epi_df class", {
+  tib <- tibble::tibble(
```
Comment (brookslogan) [inline tests/testthat/test-methods-epi_df.R:2]: `tib` sounds like it should be a tibble, but this is an `epi_df`; maybe rename to something else.

Comment (rachlobay) [reply tests/testthat/test-methods-epi_df.R]: Done.

File: R/methods-epi_df.R (hunk 2 above)
Comment (rachlobay) [reply R/methods-epi_df.R]: Added comment as to why "head" was specified in NextMethod() in file (essentially, the first arg. should be the name of the generic fun. that want to offload the work to).

File: tests/testthat/test-methods-epi_df.R (hunk 1 above)
Comment (rachlobay) [reply tests/testthat/test-methods-epi_df.R]: Sure I could add a column for that... I haven't seen it done elsewhere (ex. in the other tests). So, let me know what a reasonable example of `additional_metadata` would be and I can put that in.

**Comment by dajmcdon** [comment]
@rachlobay I spent a bit of time looking into this, and I think we're doing this the wrong way (my fault). `head()` actually calls `utils:::head.matrix()` eventually, which uses `[.tbl_df`. So rather than implementing `head.epi_df()` we should actually be implementing `[.epi_df`. Those changes will then propagate to a call to `head()` or `tail()` as well as keeping the class if someone does `some_epi_df[1:5, ]`.

I think we want it to *drop* the class and the metadata if they do column selection: `some_epi_df[ ,1:2]` should just go to a `tbl_df`.

Good examples are in [tibble](https://github.com/tidyverse/tibble/blob/main/R/subsetting.R) or probably even better [tsibble](https://github.com/tidyverts/tsibble/blob/main/R/subset.R). I suspect that we essentially need most of the code in the tSibble link.

**Review by dajmcdon — CHANGES_REQUESTED** [review]
I think there are a few additional tests that we need here, otherwise, the rest is linting.

File: R/methods-epi_df.R [hunk 4]
```diff
@@ -62,20 +62,56 @@ summary.epi_df = function(object, ...) {
...
+  not_epi_df <- !("time_value" %in% cn) || !("geo_value" %in% cn) || vctrs::vec_size(res) > nr || any(i > nr)
+  if (not_epi_df) return(tibble::as_tibble(res))
+  
+  # Case when i is numeric and there are duplicate values in it
+  if (is.numeric(i) && vctrs::vec_duplicate_any(i) > 0) return(tibble::as_tibble(res))
+  
+  # Column subsetting only, then return res as tibble
+  if (rlang::is_null(i) && !rlang::is_null(j)) return(tibble::as_tibble(res))
+  
+  att_x = attributes(x)
```
[the remark this answers is not in the document]
Comment (rachlobay) [reply R/methods-epi_df.R:110]: Done.

File: tests/testthat/test-methods-epi_df.R [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    x = 1:10,
+    y = 1:10,
+    time_value = rep(seq(
+      as.Date("2020-01-01"),
+      by = 1,
+      length.out = 5
+    ), times = 2),
+    geo_value = rep(c("ca", "hi"), each = 5)
+  ) %>% epiprocess::as_epi_df()
+
```
[the remark this answers is not in the document]
Comment (rachlobay) [reply tests/testthat/test-methods-epi_df.R:20]: Added various tests

**Review by dajmcdon — APPROVED** [review]

**Comment by brookslogan** [comment]
Some notes:
- Above test failures are related to the `additional_metadata`; at least some are due to a missing `$additional_metadata` in the tests.
- (Optional) The `additional_metadata` can be set directly in `as_epi_df`; see `pkgload::dev_help("as_epi_df")`.
- Either `additional_key_name` and `"additional_key_type"` should be changed to `other_keys` and `"<some column that acts as part of the key in addition to geo and time values>"` or turned into something without "key" in the name.  E.g., `source_url` -> `"https://example.com/data.csv"` or something like that.

File: tests/testthat/test-methods-epi_df.R (hunk 1 above)
Comment (brookslogan) [reply tests/testthat/test-methods-epi_df.R]: See `other_keys` / `source_url` example below.  Marking this resolved so just that comment remains to look at.

**Pull request merged**